smslibrary/src/main/java/com/eis/smslibrary/SMSMessage.java

```diff
@@ -38,16 +46,43 @@ public SMSMessage(@NonNull SMSPeer peer, @NonNull String messageText) throws Inv
     }
 
     /**
-     * Checks if the message content could be valid.
+     * Checks if the message content is valid.
      *
      * @param messageText to be checked.
-     * @return The state of the message after the validity tests.
+     * @return The state of the message after validity tests.
      */
-    public static ContentState checkMessageText(@NonNull String messageText) {
-        if (messageText.length() > SMSMessage.MAX_MSG_TEXT_LEN) {
-            return ContentState.MESSAGE_TEXT_TOO_LONG;
+    static ContentState checkMessageText(@NonNull String messageText) {
+        /* Since applications using this library will probably try to avoid wasting characters,
+         * they'll mostly stick to using the regular GSM character set, so we perform pattern
+         * matching on the whole message first, to avoid multiple checks on single characters
+         * when not necessary.
+         */
+        if (messageText.matches(GSM_CHARACTERS_STRING_REGEX)) {
+            // messageText contains only GSM characters
+            if (messageText.length() <= MAX_MSG_TEXT_LEN) {
+                return ContentState.MESSAGE_TEXT_VALID;
+            }
+        } else if (!messageText.matches(GSM_CHARACTERS_EXTENSION_STRING_REGEX)) {
```

Comment:
if is utf 8 android sends an mms

Reply:
Did you test this using this library? The [documentation](https://developer.android.com/reference/android/telephony/SmsManager.html#sendMultipartTextMessage(java.lang.String,%20java.lang.String,%20java.util.ArrayList%3Cjava.lang.String%3E,%20java.util.ArrayList%3Candroid.app.PendingIntent%3E,%20java.util.ArrayList%3Candroid.app.PendingIntent%3E)) of method `sendMultiPartTextMessage` doesn't mention MMS.

Reply:
i know because i was sending a number casted to char a message was cutted to 70(non utf-8 char) and a message(utf-8) was sended but i dont received it in the other simulator
i think it has become an mms
i'm not sure but you can't catch it in the onReceive method of the broadcast

Reply:
in my opinion is better to forbid every non gms standard alphabet message

Reply:
I don't think the encoding used is ever UTF-8, because whenever you send a message that includes non-GSM characters they're all encoded in 16 bits each

Reply:
i think(not 100% sure) android encodes utf-16 when it finds a non utf-8 char

Reply:
Could you point me to the code of the app you used to test sending of a number casted to `char`? Otherwise I'll write a simple test app to see if what you described happens with this library too.

Reply:
if you want you can try sending messages between 2 simulators cause it works in android api and you can't reach it with tests

Reply:
Ok, I'll try it as soon as possible

Reply:
So I tried sending messages both with the test app and the default Messaging app, and to my surprise they both behaved incorrectly in the same way, I think this must be a problem of my emulator. I recorded a video of what happened: [https://www.youtube.com/watch?v=1X0zlCuF_sk](https://www.youtube.com/watch?v=1X0zlCuF_sk)

Reply:
> i think(not 100% sure) android encodes utf-16 when it finds a non utf-8 char

I read that it uses the dafault which is UTF-8 unless otherwise specified

Reply:
[I ran another test](https://www.youtube.com/watch?v=QlzJoHAsAa4), this time with two emulators, as you can see messages are split in multiple messages both when sending them with our app and the emulator, the funny thing is that the message sent with our app was received in its entirety, while the message sent with the default SMS app was not completely received.

Reply:
[Here](https://github.com/EIS0/smslibrary/tree/feature/test-app)'s the code of the app I wrote to test the library, in case anybody wants to test multi-part messages on their emulators

Reply:
but why not entirely?

Reply:
I have no idea, especially because there shouldn't be any network issues between emulators. If you could test this app in your emulators you'd do me a favor.

Reply:
By the way, now that I think about it this issue is probably caused by a bad implementation of the emulator's virtual network. In the first weeks of the development of our group's library, I tried testing delivery confirmation of texts, and while it worked on my phone it didn't work in emulators, I wouldn't receive the delivery confirmation even minutes after the message was received on the other emulator. @d-u-d-e tomorrow I'll send you a multi-part SMS using this library on my phone, and we'll see if it's received correctly.

Reply:
I tested multi part message sending on my phone with @d-u-d-e and it works, I can't publish the video where I send the messages using the app because Mariotto's number is shown, but I sent it to him and he can confirm that they're received correctly, both when using standard GSM characters and characters not included in the GSM charset, like emojis.

Reply:
We did this test between two physical phones, and concluded that messages are sent correctly. This issue might be caused by a bad implementation of the emulators virtual network.